dbms/src/AggregateFunctions/AggregateFunctionSumMap.h

```diff
@@ -253,6 +258,47 @@ class AggregateFunctionSumMap final : public IAggregateFunctionDataHelper<
     }
 
     const char * getHeaderFilePath() const override { return __FILE__; }
+
+    virtual bool keepKey(const T & key) const = 0;
+};
+
+template <typename T>
+class AggregateFunctionSumMap final : public AggregateFunctionSumMapBase<T, AggregateFunctionSumMap<T>>
+{
+public:
+    AggregateFunctionSumMap(const DataTypePtr & keys_type, DataTypes & values_types)
+        : AggregateFunctionSumMapBase<T, AggregateFunctionSumMap<T>>{keys_type, values_types}
+    {}
+
+    String getName() const override { return "sumMap"; }
+
+    bool keepKey(const T &) const override { return true; }
+};
+
+template <typename T>
+class AggregateFunctionSumMapFiltered final : public AggregateFunctionSumMapBase<T, AggregateFunctionSumMapFiltered<T>>
+{
+private:
+    std::vector<T> keys_to_keep;
+
+public:
+    AggregateFunctionSumMapFiltered(const DataTypePtr & keys_type, const DataTypes & values_types, const Array & keys_to_keep)
+        : AggregateFunctionSumMapBase<T, AggregateFunctionSumMapFiltered<T>>{keys_type, values_types}
+    {
+        this->keys_to_keep.reserve(keys_to_keep.size());
+        for (const Field & f : keys_to_keep)
+        {
+            this->keys_to_keep.emplace_back(f.safeGet<NearestFieldType<T>>());
+        }
+        std::sort(begin(this->keys_to_keep), end(this->keys_to_keep));
+    }
+
+    String getName() const override { return "sumMapFiltered"; }
+
+    bool keepKey(const T & key) const override
+    {
+        return std::binary_search(begin(keys_to_keep), end(keys_to_keep), key);
```

Comment:
nan values can confuse binary search.

Reply:
A hash table (even `std::unordered_map`) should be better.

Reply:
I didn't know about the NaN thing an `binary_search`, thanks! I went for an `unordered_set` but had to implement `std::hash` specializations for `Decimal<T>`.